docs(output-targets): update dist-custom-elements docs (#925)
ionic-team/stencil-site: src/docs/output-targets/custom-elements.md

  - adamdbradley
  - rwaskiewicz
  - splitinfinities
  - tanner-reits
---

# Custom Elements

> The `dist-custom-elements` output target is used to generate custom elements in a more optimized way for tree shaking, and it's the recommended approach when using any frontend framework integrations.

Reviewer: The "in a more optimized way for tree shaking" was originally written (albeit unclearly) to differentiate it from `dist-custom-elements-bundle`. Since we'll be removing that in the near future, I wonder if we should doctor this sentence a little bit to either
- compare it against `www`
- talk about it a little more generally - e.g. "what's it do?"

Author: I think we kind of address comparing it to `www` with the other callout you had for lazy-loading vs. non lazy-loading (since I feel like that's the big difference most users would make a decision against). 

What else would we generally talk about? Elaborate more on treeshaking? Talk about using Rollup under-the-hood and what that generates? In that case, we might just keep what was originally there but wordsmith it a bit

Reviewer: I think that currently (and before this PR) we didn't really do a good job at giving a high level description of this output target. It doesn't feel we didn't really answer "What is this?", rather we compared it to `dist-custom-elements` and said it was recommended for frontend integrations. I don't think I'm describing it well, but it feels like there's an executive summary/elevator pitch for this output target that could be added here. Does that make sense?

Author: Alright, added a few sentences that I morphed from a few other doc pages that were doing some of these comparisons

This target can be used outside of frameworks as well, if lazy-loading functionality is not required or desired.

```tsx
// stencil.config.ts
import { Config } from '@stencil/core';

export const config: Config = {
  // Other top-level config options here
  outputTargets: [
    {
      type: 'dist-custom-elements',
      // Output target config options here
    },
    // Other output targets here
  ],
};
```

## Config

### copy

_default: `undefined`_

An array of [copy tasks](/docs/copy-tasks) to be executed during the build process.

### customElementsExportBehavior

_default: `'default'`_

> This config option provides additional behaviors that will alter the default component export _OR_ custom element definition behaviors

Author: I like that 👍

  - adamdbradley
  - rwaskiewicz
  - splitinfinities
  - tanner-reits
---

# Custom Elements

> The `dist-custom-elements` output target is used to generate custom elements in a more optimized way for tree shaking, and it's the recommended approach when using any frontend framework integrations. This output target creates custom elements that directly extend `HTMLElement` and provides simple utility functions for easily defining these elements on the [Custom Element Registry](https://developer.mozilla.org/en-US/docs/Web/API/CustomElementRegistry). As stated, this output target excels in use in frontend frameworks and projects that will handle bundling, lazy-loading, and custom element registration themselves.

Reviewer: I like it - one reword suggestion to drop some of the duplication:
```suggestion
The `dist-custom-elements` output target creates custom elements that directly extend `HTMLElement` and provides simple utility functions for easily defining these elements on the [Custom Element Registry](https://developer.mozilla.org/en-US/docs/Web/API/CustomElementRegistry). This output target excels in use in frontend frameworks and projects that will handle bundling, lazy-loading, and custom element registration themselves.
```

This target can be used outside of frameworks as well, if lazy-loading functionality is not required or desired. For using lazily loaded Stencil components, please refer to the [www output target](/docs/www).

To generate components using the `dist-custom-elements` output target, add it to a project's `stencil.config.ts` file like so:

```tsx
// stencil.config.ts
import { Config } from '@stencil/core';

export const config: Config = {
  // Other top-level config options here
  outputTargets: [
    {
      type: 'dist-custom-elements',
      // Output target config options here
    },
    // Other output targets here
  ],
};
```

## Config

### copy

_default: `undefined`_

An array of [copy tasks](/docs/copy-tasks) to be executed during the build process.

### customElementsExportBehavior

_default: `'default'`_

By default, the `dist-custom-elements` output target generates a single file per component, and exports each of those files individually.

> In some cases, library authors may want to change this behavior, either to automatically define component children, provide a single file containing all component exports, etc.

Reviewer 2: ```suggestion
In some cases, library authors may want to change this behavior, for instance to automatically define component children, provide a single file containing all component exports, etc.
```

| `auto-define-custom-elements` | A component and its children will be automatically defined with the `CustomElementRegistry` when the component's module is imported.                                                                                                                                                                                                                                                                      |
| `single-export-module`        | All component and custom element definition helper functions will be exported from the `index.js` file in the output directory (see [Defining Exported Custom Elements](#defining-exported-custom-elements) for more information on this file's purpose). This file can be used as the root module when distributing your component library, see [below](#distributing-custom-elements) for more details. |

> > Note: At this time, components created not using JSX may not be automatically defined. This is a known limitation of the API and users should be aware of it.

Author: Looks like @alicewriteswrongs has a PR open to change this block

> > Note: At this time, components created not using JSX may not be automatically defined. This is a known limitation of the API and users should be aware of it.

Reviewer 2: I just thought it sounded a bit odd!

> > Note: At this time, components created not using JSX may not be automatically defined. This is a known limitation of the API and users should be aware of it.

Reviewer: It did. My bad 🙃

> > Note: At this time, components created not using JSX may not be automatically defined. This is a known limitation of the API and users should be aware of it.

Reviewer 2: Since this blockquote has now been changed on `main` should we merge `main` into the v3 branch and rebase this PR over that to make sure things are consistent? I think it would be easy to lose a change later in a merge commit somewhere.

> > Note: At this time, components created not using JSX may not be automatically defined. This is a known limitation of the API and users should be aware of it.

Reviewer: Yah - I'll do that after we merge this into v3 (there're already conflicts there, and I'd like to do them in one fell swoop)

### dir

_default: `'dist/components'`_

This config option allows you to change the output directory where the compiled output for this output target will be written.

### empty

_default: `true`_

If `true`, this config option will remove the contents of the [output directory](#dir) before writing the new build output.

### generateTypeDeclarations

_default: `true`_

By default, Stencil will generate type declaration files (`.d.ts` files) as a part of the `dist-custom-elements` output target through the `generateTypeDeclarations` field on the target options. Type declaration files will be placed in the `dist/types` directory in the root of a Stencil project. At this time, this output destination is not able to be configured.

Setting this flag to `false` will not generate type declaration files for the `dist-custom-elements` output target.

> > When set to generate type declarations, Stencil respects the export behavior selected via `customElementsExportBehavior` and generates type declarations specific to the content of that file.

Author: Yeah this is weird cuz really what it does is generate type declarations for the `component.d.ts` file and then just re-export the types from `dist/types`. Changed this up a bit to call out what we were really trying to call attention to

### inlineDynamicImports

_default: `false`_

> Setting this flag to `true` will trigger inline dynamic imports for the under-the-hood Rollup bundle output. For more information, see [Rollup's documentation](https://rollupjs.org/guide/en/#outputinlinedynamicimports) for this option.

Author: Think I got it in layman's terms

Also, I think there might be an issue with setting this property on the target rn... was seeing some build errors when playing around with it

### includeGlobalScripts

_default: `false`_

> Setting this flag to `true` will include global scripts in the bundle and execute them once the bungle entry point in loaded.

Author: Oh maybe I misunderstood this option. I thought it would point to https://stenciljs.com/docs/config#globalscript

> Setting this flag to `true` will include global scripts in the bundle and execute them once the bungle entry point in loaded.

Reviewer: Hmmm maybe I'm wrong? Now I'm second guessing myself 😆

> Setting this flag to `true` will include global scripts in the bundle and execute them once the bungle entry point in loaded.

Author: It does refer to what I linked, just a heads up

> 

Author: Omitted `minify` option here. I'll add that to my `externalRuntime` PR since it relates to some changes there

Author: Yeah! I'll cherry-pick over the commit from the other PR that I want to have for this to reference

| `single-export-module`        | All component and custom element definition helper functions will be exported from the `index.js` file in the output directory (see [Defining Exported Custom Elements](#defining-exported-custom-elements) for more information on this file's purpose). This file can be used as the root module when distributing your component library, see [below](#distributing-custom-elements) for more details. |

> Note: At this time, components created not using JSX may not be automatically defined. This is a known limitation of the API and users should be aware of it.

### dir

_default: `'dist/components'`_

This config option allows you to change the output directory where the compiled output for this output target will be written.

### empty

_default: `true`_

> If `true`, this config option will remove the contents of the [output directory](#dir) between builds.

Reviewer 2: what about phrasing this (and similar ones) more like

```suggestion
If `true`, the contents of the [output directory](#dir) will be removed between builds.
```

Author: Went a slightly different route, but did change a few phrases so they were consistent across the options

### externalRuntime

_default: `true`_

When `true`, this flag results in the following behaviors:

1. Minification will follow what is specified in the [Stencil config](docs/config#minifyJs).
2. Filenames will not be hashed
3. All imports from packages under `@stencil/core/*` will be marked as external and therefore not included in the generated Rollup bundle.

### generateTypeDeclarations

_default: `true`_

By default, Stencil will generate type declaration files (`.d.ts` files) as a part of the `dist-custom-elements` output target through the `generateTypeDeclarations` field on the target options. Type declaration files will be placed in the `dist/types` directory in the root of a Stencil project. At this time, this output destination is not able to be configured.

Setting this flag to `false` will not generate type declaration files for the `dist-custom-elements` output target.

> > When set to generate type declarations, Stencil respects the export behavior selected via `customElementsExportBehavior` and generates type declarations specific to the content of the generated `component.d.ts` file.

Reviewer 2: is it specific to the generated `component.d.ts` or rather that the types that it will generate for `component.d.ts` are appropriate for what is exported given the `customElementsExportBehavior` ? Maybe I'm not understanding something correctly here.

Author: Yeah I did make this confusing and botched it a bit. Pushed a change to hopefully make that clearer

bundling, and where your assets can be loaded from. How the files are copied to the production build directory depends on the bundler or tooling.
The configs below provide examples of how to do this automatically with popular bundlers.

> ## Consuming Custom Elements

Reviewer: I'd consider putting this above the "Making Assets Available". I have a feeling this will get more views than that section (but I could be wrong 🤷)

By default, the custom elements files will be written to `dist/components/`. This directory can be configured using the output target's `dir` config.

> The generated files will each export a component class and will already have the styles bundled. However, this build does not define the custom elements or apply any polyfills. Any dependencies of your imported component will need to be loaded as well.

Author: I took that out and added clarification for assets. Am a bit confused why this was here in the first place. Dependencies are bundled, are they not? And child components are included in a parent's `defineCustomElement` function.

}
```

> Note: If you set the `dir` property on the output target config, replace `dist/components` in the above snippet with the path set in the config.

> ## Example Bundler Configs

Author: Left this alone. I'm not too sure of it's helpfulness, but it looks like it was added ~2 years ago by a community member. Long term, I'd even consider use separating these output target docs into two pages: one for setting up/configuring, and one for consuming the output in a project.

};
```

> ## How is this different from the "dist" and the "dist-custom-element-bundle" output targets?

Author: Killed this since bundle is going away in v3

};
```

> <!-- TODO: remove once custom elements bundle is officially not supported -->

Reviewer 2: should we put a JIRA ticket number in here for easy grepping later?
